File: python/2671. Frequency Tracker.py
class FrequencyTracker:

    def __init__(self):
        self.num_freq = [0] * 100001
        self.freq_freq = [0] * 100001

    def add(self, number: int) -> None:
        if self.num_freq[number] > 0:
            self.freq_freq[self.num_freq[number]] -= 1
        self.num_freq[number] += 1
        self.freq_freq[self.num_freq[number]] += 1

    def deleteOne(self, number: int) -> None:
        if self.num_freq[number] == 0: return
        self.freq_freq[self.num_freq[number]] -= 1
        self.num_freq[number] -= 1
        self.freq_freq[self.num_freq[number]] += 1

    def hasFrequency(self, frequency: int) -> bool:
        return self.freq_freq[frequency] > 0


# Your FrequencyTracker object will be instantiated and called as such:
# obj = FrequencyTracker()
# obj.add(number)
# obj.deleteOne(number)
# param_3 = obj.hasFrequency(frequency)

File: java/2671. Frequency Tracker.java
class FrequencyTracker {

    private int[] numFreq;
    private int[] freqFreq;

    public FrequencyTracker() {
        numFreq = new int[100001];
        freqFreq = new int[100001];
    }

    public void add(int number) {
        int freq = numFreq[number]++;
        if (freq > 0) freqFreq[freq]--;
        freqFreq[freq+1]++;
    }

    public void deleteOne(int number) {
        if (numFreq[number] == 0) return;
        int freq = numFreq[number]--;
        freqFreq[freq]--;
        if (freq > 0) freqFreq[freq-1]++;
    }

    public boolean hasFrequency(int frequency) {
        return freqFreq[frequency] > 0;
    }
}

/**
 * Your FrequencyTracker object will be instantiated and called as such:
 * FrequencyTracker obj = new FrequencyTracker();
 * obj.add(number);
 * obj.deleteOne(number);
 * boolean param_3 = obj.hasFrequency(frequency);
 */
